proposals/extensions.md

```diff
@@ -1219,34 +1219,68 @@ implicit extension E for int
-into the ref struct and a constructor.  
-TODO2(instance) The wrapping can be done with a static unspeakable factory method  
+## Instance field
 
-Values of extension types are left as values of the underlying value type, until an extension
-member is accessed. When an extension member is accessed, an extension instance is created
-with a reference to the underlying value and the member is accessed on that instance.
+If the extension not static, then we emit a private instance field of the underlying type
+into the struct.  
+We use the default layout for structs (sequential layout, with pack and size 0)
+ensuring that the single instance field is placed at offset zero.
 
+Note: Although we could not find an explicit statement to that effect,
+this behavior falls out from ECMA 335 (II.10.1.2), which states for sequential layout:
+> The CLI shall lay out the fields in sequential order, based on the order of the fields in the logical metadata table (§II.22.15).
+
+> [Rationale: ... sequential layout is intended to instruct the CLI to match layout rules commonly followed by languages like C and C++
+> on an individual platform, where this is possible while still guaranteeing verifiable layout. ...]
+
+and from the C99 standard section 6.7.2.1 bullet point 13 (gated):
+
+> Within a structure object, the non-bit-field members and the units in which bit-fields 
+> reside have addresses that increase in the order in which they are declared. 
+> A pointer to a structure object, suitably converted, points to its initial member 
+> (or if that member is a bit-field, then to the unit in which it resides), 
+> and vice versa. There may be unnamed padding within a structure object, but not at its beginning.
+
+For example `implicit extension E for UnderlyingType` yields
+```csharp
+struct E
+{
+	private UnderlyingType <UnderlyingInstance>$;
+	private static void <ImplicitExtension>$(UnderlyingType) { }
+}
 ```
-Extension r = default(UnderlyingType); // emitted as a local of type `UnderlyingType`
-r.ExtensionMember(); // emitted as `new Extension(ref r).ExtensionMember();`
-```
+
+## Instance invocations
+
+The design and layout of the instance field will allow us to re-interpret an instance of `UnderlyingType`
+as an instance of `E` with `public static ref TTo Unsafe.As<TFrom,TTo>(ref TFrom source)`.  
+
+If the receiver of an extension member invocation on a value of the underlying type is `r`,
+we will replace it with `Unsafe.As<UnderlyingType, E>(ref r)`.
+If `r` does not refer to a location, a temporary variable will be created and initialized, and its reference
+will be used.
+
+The re-interpreted receiver will be only computed once where possible. For example, in a compound assignment
+`r.P += value;`.
+
+TODO we may wrap this method in a compiler-generated helper to increase verifiability of the generated code.  
+TODO there is also a verifiability issue with taking a reference to `this` (which is readonly)
```

Comment:
There is a helper that allows to drop the `readonly` property of a `ref` (Unsafe.AsRef<T>), but in some scenarios I think we should capture `readonly` targets by value

Reply:
It would be great if there was a way to specify whether a reference is wanted or not for a struct extension's member. My understanding is that a number of VectorXxx members are extensions / static methods / operators because they can be passed by value instead of reference. And I've seen similar for enums too, along with other primitives. But there are also many scenarios where it's great to pass by reference instead (like calling an mutating interface member in a constrained way, or any other mutating method, or for very large structs). Don't really see how this could work with the current design though, unless we lowered these special (by value) instance methods/members as static methods with extra first parameter of extension type and some name prefix/suffix to indicate to the compiler that they're instance methods, but by value instead.